### PR TITLE
average transaction fee for block added

### PR DESCRIPTION
**Pull request by waqarRns**

### CODE REVIEW

File: src/modules/storage/LedgerStorage.ts [hunk 1]
```diff
@@ -357,6 +369,57 @@ export class LedgerStorage extends Storages
...
+    /**
+     * Saving Average Fees
+     */
+    public putfees (block: Block) {
+        function save_fee (storage: LedgerStorage, height: Height, time_stamp: number, average_tx_fee: JSBI, total_tx_fee: JSBI, total_payload_fee: JSBI, total_fee: JSBI) {
+            return new Promise<void>((resolve, reject) => {
+                storage.run(
+                    `INSERT INTO fees
+                    (height, time_stamp, average_tx_fee, total_tx_fee, total_payload_fee, total_fee)
+                VALUES
```
Comment (TrustHenry) [inline src/modules/storage/LedgerStorage.ts]:
```suggestion
                    VALUES
```

File: src/modules/storage/LedgerStorage.ts [hunk 2]
```diff
@@ -357,6 +369,57 @@ export class LedgerStorage extends Storages
...
+                        total_fee.toString(),
+                    ]
+                )
+                    .then(() => {
+                        resolve();
+                    })
+                    .catch((err) => {
+                        reject(err);
+                    })
+            })
```
Comment (TrustHenry) [inline src/modules/storage/LedgerStorage.ts]:
```suggestion
                    });
            });
```

File: tests/Boascan.test.ts [hunk 3]
```diff
@@ -597,4 +597,24 @@ describe('Test of Stoa API Server', () => {
...
+                    time_stamp: 1609459200, 
+                    average_tx_fee: 0 
+                },
+                { 
+                    height: 1, 
+                    time_stamp: 1609459800, 
+                    average_tx_fee: 0 
+                }
+        ]
+        assert.deepStrictEqual(response.data, expected)
```
Comment (TrustHenry) [inline tests/Boascan.test.ts]:
```suggestion
        assert.deepStrictEqual(response.data, expected);
```

File: src/modules/storage/LedgerStorage.ts [hunk 4]
```diff
@@ -2257,6 +2320,22 @@ export class LedgerStorage extends Storages
...
+    */
+    public calculateAvgFeeChart (from: number, to: number): Promise<any[]> {
+        let sql =
+            `SELECT
+                height, time_stamp, average_tx_fee
+             FROM
+                fees
+             WHERE
+                time_stamp BETWEEN ? AND  ? `;
+        return this.query(sql, [from, to])
```
Comment (TrustHenry) [inline src/modules/storage/LedgerStorage.ts]:
```suggestion
        return this.query(sql, [from, to]);
```

File: src/Stoa.ts [hunk 5]
```diff
@@ -155,6 +155,7 @@ class Stoa extends WebService
         this.app.post("/block_externalized", this.postBlock.bind(this));
         this.app.post("/preimage_received", this.putPreImage.bind(this));
         this.app.post("/transaction_received", this.putTransaction.bind(this));
+        this.app.get("/averagefeechart", this.averageFeeChart.bind(this));
```
Comment (TrustHenry) [inline src/Stoa.ts]:
```suggestion
        this.app.get("/average_fee_chart", this.averageFeeChart.bind(this));
```

File: src/Stoa.ts [hunk 6]
```diff
@@ -1805,6 +1806,69 @@ class Stoa extends WebService
             })
     }
 
+    /**
+     * GET /averagefeechart/
```
Comment (TrustHenry) [inline src/Stoa.ts]:
```suggestion
     * GET /average_fee_chart/
```

File: src/Stoa.ts [hunk 7]
```diff
@@ -1805,6 +1806,69 @@ class Stoa extends WebService
...
+     * Called when a request is received through the `/averagefeechart/` handler
```
Comment (TrustHenry) [inline src/Stoa.ts]:
```suggestion
     * Called when a request is received through the `/average_fee_chart/` handler
```

File: src/Stoa.ts [hunk 8]
```diff
@@ -1805,6 +1806,69 @@ class Stoa extends WebService
...
+    {
+        let filter_end = await Time.msToTime(Date.now());
+        let filter_begin: number;
+        if (req.query.beginDate === undefined)
+        {
+            res.status(400).send(`Parameter beginDate must also be set.`);
+            return;
+        }
+
+        logger.http(`GET /averagefeechart/${req.query.beginDate.toString()}`);
```
Comment (TrustHenry) [inline src/Stoa.ts]:
```suggestion
        logger.http(`GET /average_fee_chart/${req.query.beginDate.toString()}`);
```

File: docs/Database_Schema.md [hunk 9]
```diff
@@ -467,3 +467,28 @@ CREATE TABLE IF NOT EXISTS tx_pool (
...
+```sql
+CREATE TABLE IF NOT EXISTS "fees"(
+    "height"             INTEGER    NOT NULL,
+    "time_stamp"         INTEGER    NOT NULL,
+    "average_tx_fee"     BIGINT(20) NOT NULL,
+    "total_tx_fee"       BIGINT(20) NOT NULL,
+    "total_payload_fee"  BIGINT(20) NOT NULL,
+    "total_fee"          BIGINT(20) NOT NULL,
+    PRIMARY KEY(height)
+);
```
Comment (TrustHenry) [inline docs/Database_Schema.md:546]: CREATE INDEX ind_time_stamp on fees(time_stamp);

File: src/modules/storage/LedgerStorage.ts [hunk 10]
```diff
@@ -2257,6 +2320,22 @@ export class LedgerStorage extends Storages
...
+    * @param to End date for chart history
+    */
+    public calculateAvgFeeChart (from: number, to: number): Promise<any[]> {
+        let sql =
+            `SELECT
+                height, time_stamp, average_tx_fee
+             FROM
+                fees
+             WHERE
+                time_stamp BETWEEN ? AND  ? `;
```
Comment (TrustHenry) [inline src/modules/storage/LedgerStorage.ts]: When querying a wide range (annual) of data, a lot of data can occur.

**Review by TrustHenry — CHANGES_REQUESTED** [review]

File: src/modules/storage/LedgerStorage.ts (hunk 10 above)
Comment (waqarRns) [reply src/modules/storage/LedgerStorage.ts]: we have to response with whole data within 
range. i don't think pagination is needed here.

File: src/Types.ts [hunk 11]
```diff
@@ -811,3 +811,39 @@ export interface IMarketChart
...
+    average_tx_fee: number;
+
+    /**
+     * Total Fee of Transactions in Block
+     */
+    total_tx_fee: number;
+
+    /**
+      * Total Payload Fee
+      */
```
Comment (TrustHenry) [inline src/Types.ts]:
```suggestion
     * Total Payload Fee
     */
```

File: src/modules/storage/LedgerStorage.ts [hunk 12]
```diff
@@ -357,6 +370,57 @@ export class LedgerStorage extends Storages
             })();
         });
     }
+    /**
```
Comment (TrustHenry) [inline src/modules/storage/LedgerStorage.ts:398]:
```suggestion
    
    /**
```

**Comment by TrustHenry** [comment]
Conflicting files
docs/Database_Schema.md
src/Stoa.ts
src/modules/storage/LedgerStorage.ts
tests/Boascan.test.ts

File: src/modules/storage/LedgerStorage.ts (hunk 10 above)
Comment (TrustHenry) [reply src/modules/storage/LedgerStorage.ts]: I think there is enough room for improvement here.
Please prepare improvement PR next time.

File: tests/Boascan.test.ts [hunk 13]
```diff
@@ -640,4 +640,25 @@ describe("Test of Stoa API Server", () => {
...
+//     { height: 3, time_stamp: 1609461000, average_tx_fee: 540 },
+//     { height: 4, time_stamp: 1609461600, average_tx_fee: 540 },
+//     { height: 5, time_stamp: 1609462200, average_tx_fee: 540 },
+//     { height: 6, time_stamp: 1609462800, average_tx_fee: 540 },
+//     { height: 7, time_stamp: 1609463400, average_tx_fee: 540 },
+//     { height: 8, time_stamp: 1609464000, average_tx_fee: 540 },
+//     { height: 9, time_stamp: 1609464600, average_tx_fee: 540 }
+//   ]
+//         assert.deepStrictEqual(response.data, expected);
+    });
```
Comment (TrustHenry) [inline tests/Boascan.test.ts]: The test code is not ready yet.

**Comment by TrustHenry** [comment]
Can you resolve the conflict and make it one commit?

File: src/Stoa.ts [hunk 14]
```diff
@@ -2020,6 +2023,115 @@ class Stoa extends WebService {
             })
     }
 
+    /**
+     * GET /average_fee_chart/
```
Comment (TrustHenry) [inline src/Stoa.ts]:
```suggestion
     * GET /average_fee_chart
```

File: src/Stoa.ts [hunk 15]
```diff
@@ -2020,6 +2023,115 @@ class Stoa extends WebService {
...
+     * Called when a request is received through the `/average_fee_chart/` handler
+     * The parameter `beginDate` is the start date of the range of dates to look up.
+     * @returns Returns average transaction fee between range ( beginDate - currentDate )
```
Comment (TrustHenry) [inline src/Stoa.ts]: The document needs to be updated.

File: src/Stoa.ts [hunk 16]
```diff
@@ -2020,6 +2023,115 @@ class Stoa extends WebService {
...
+        if (req.query.filter === undefined) {
+            res.status(400).send(`Parameter filter must also be set.`);
+            return;
+        }
+        else if(req.query.date === undefined)
+        {
+            res.status(400).send(`Parameter endDate must also be set.`);
+            return;
+        }
+        else if ( req.query.date !== undefined) {
```
Comment (TrustHenry) [inline src/Stoa.ts]:
```suggestion
        else {
```

File: src/Stoa.ts [hunk 17]
```diff
@@ -2020,6 +2023,115 @@ class Stoa extends WebService {
...
+                                total_payload_fee: row.total_payload_fee,
+                                total_fee: row.total_fee
+                            }
+                        )
+                    }
+                    return res.status(200).send(JSON.stringify(avgFeelist));
+                }
+            })
+            .catch((err) => {
+                logger.error("Failed to data lookup to the DB: " + err,
```
Comment (TrustHenry) [inline src/Stoa.ts]:
```suggestion
                logger.error("Failed to averageFeeChart data lookup to the DB: " + err,
```

**Review by TrustHenry — APPROVED** [review]
LGTM

**Pull request merged**